### PR TITLE
Add support for PCNC1 board (as used in SkyViper)

### PR DESCRIPTION
**Pull request by peterbarker**
These patches are *on top of* another PR, which adds support for the ICM20789 to ArduPilot.

Each of the relevant commits for *this* PR have PCNC1 in the name.

This is not a complete patch set - there are more things to come.  But these patches do allow you to build something which boots and provides mavlink - and thus allows testing of the ICM20789 driver.

Due to parameter differences this won't work directly on a SkyViper - SR1_BAUDRATE will need to be reset in the ROMFS so the Sonix can talk to the board.

### CODE REVIEW

**Comment by tridge** [comment]
I don't think we need the --board target, we only need the board number in the enum, plus the baro, compass and ins switch changes. Setting the BRD_TYPE can be done with apj_tool, along with all the other needed parameters

File: libraries/AP_Baro/AP_Baro_ICM20789.cpp [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+        if (!send_cmd16(0xC7F7)) {
+            debug("ICM20789: read cal2[%u] failed\n", i);
+            return false;
+        }
+        uint8_t d[3];
+        if (!dev->transfer(nullptr, 0, d, sizeof(d))) {
+            debug("ICM20789: read cal3[%u] failed\n", i);
+            return false;
+        }
+        sensor_constants[i] = int16_t((d[0]<<8) | d[1]);
```
Comment (muramura) [inline libraries/AP_Baro/AP_Baro_ICM20789.cpp]: It reads 3 bytes and uses 2 bytes. The remaining 1 byte?

File: libraries/AP_Baro/AP_Baro_ICM20789.cpp [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+        _sem->give();
+    }
+}
+
+void AP_Baro_ICM20789::timer(void)
+{
+    uint8_t d[9] {};
+    if (dev->transfer(nullptr, 0, d, sizeof(d))) {
+        // ignore CRC bytes for now
+        uint32_t Praw = (uint32_t(d[0]) << 16) | (uint32_t(d[1]) << 8) | d[3];
```
Comment (muramura) [inline libraries/AP_Baro/AP_Baro_ICM20789.cpp]: Is not the pressure value a continuous byte? d [2]?

File: libraries/AP_Baro/examples/ICM20789/ICM20789.cpp [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+    spi_dev->set_read_flag(0x80);
+
+    // run the SPI bus at low speed
+    spi_dev->set_speed(AP_HAL::Device::SPEED_LOW);
+
+    uint8_t whoami = 0;
+    uint8_t v;
+
+    spi_dev->write_register(0x6B, 0x01);
+    spi_dev->write_register(0x6B, 0x01);
```
Comment (muramura) [inline libraries/AP_Baro/examples/ICM20789/ICM20789.cpp]: Is it fun to set the same value in the same register?

**Comment by peterbarker** [comment]
On Fri, 5 Jan 2018, Andrew Tridgell wrote:

> I don't think we need the --board target, we only need the board number in
> the enum, plus the baro, compass and ins switch changes. Setting the
> BRD_TYPE can be done with apj_tool, along with all the other needed
> parameters

FLASH_STORAGE is switched based on the board type.  That's the only reason 
I added --board

Your changes to make that run-time aren't written yet AIUI...

**Comment by peterbarker** [comment]
On Fri, 5 Jan 2018, murata,katsutoshi wrote:

> Since there are places that do the same kind of processing, if you change,
> other processing is also required. It's confirmed. In programming education
> in Japan, we are advancing to put white space between operator and
> variable. It has the effect of raising the readability of the expression.
> Is not there such a coding rule abroad?

Rules are made to be broken :-)  Sometimes things are just clearer without 
the extra whitespace as it logically groups things together (my opinion, 
of course).

e.g.
foo(5, (a > b) ? 17 : 35)
foo(5, (a>b)?17:35)

Of course, you're probably still doing something wrong if you're writing 
this ;-)

> > @@ -0,0 +1,336 @@
> +/*
> +   This program is free software: you can redistribute it and/or modify
> 
> I got a comment from @tridge that the license notation of GPLv3 is not
> mandatory.
> If it is new, I do not need to mention it.

Interesting point.  This one is here because it is going to be of great 
interest to others seeking to support this new sensor and it is important 
the license be made clear.  (my first guess; my second guess is that some 
other driver was cp'd into place and then edited :-) )

> +bool AP_Baro_ICM20789::send_cmd16(uint16_t cmd)
> +{
> +    uint8_t cmd_b[2] = { uint8_t(cmd >> 8), uint8_t(cmd & 0xFF) };
> +    return dev->transfer(cmd_b, 2, nullptr, 0);
> 
> I think it would be better to get the number of bytes by sizeof byte array.

Good call.

> ___________________________________________________________________________
> 
> In libraries/AP_Baro/AP_Baro_ICM20789.cpp:
> 
> > +    const uint8_t cmd[5] = { 0xC5, 0x95, 0x00, 0x66, 0x9C };
> +    if (!dev->transfer(cmd, sizeof(cmd), nullptr, 0)) {
> +        debug("ICM20789: read cal1 failed\n");
> +        return false;
> +    }
> +    for (uint8_t i=0; i<4; i++) {
> +        if (!send_cmd16(0xC7F7)) {
> +            debug("ICM20789: read cal2[%u] failed\n", i);
> +            return false;
> +        }
> +        uint8_t d[3];
> +        if (!dev->transfer(nullptr, 0, d, sizeof(d))) {
> +            debug("ICM20789: read cal3[%u] failed\n", i);
> +            return false;
> +        }
> +        sensor_constants[i] = int16_t((d[0]<<8) | d[1]);
> 
> It reads 3 bytes and uses 2 bytes. The remaining 1 byte?

One byte of CRC - ignored.  Will add a comment along those lines.

> +void AP_Baro_ICM20789::timer(void)
> +{
> +    uint8_t d[9] {};
> +    if (dev->transfer(nullptr, 0, d, sizeof(d))) {
> +        // ignore CRC bytes for now
> +        uint32_t Praw = (uint32_t(d[0]) << 16) | (uint32_t(d[1]) << 8) | d
> [3];
> 
> Is not the pressure value a continuous byte? d [2]?

CRC bytes, as the comment there's getting at :-)  I'll attempt to clarify 
the comment.

> +static void spi_init()
> +{
> +    // SPI reads have flag 0x80 set
> +    spi_dev->set_read_flag(0x80);
> +
> +    // run the SPI bus at low speed
> +    spi_dev->set_speed(AP_HAL::Device::SPEED_LOW);
> +
> +    uint8_t whoami = 0;
> +    uint8_t v;
> +
> +    spi_dev->write_register(0x6B, 0x01);
> +    spi_dev->write_register(0x6B, 0x01);
> 
> Is it fun to set the same value in the same register?

Yes.  Thankfullly tridge has already done all of the hair-pulling required 
to get this sensor to work.  Including things like writing registers 
twice, getting all of the delays of the correct length, jumping through 
hoops with register settings in initialisation, ....

Thanks for the review!  I've pushed patches up on top reflecting my 
comments here.

**Comment by tridge** [comment]
this has been merged via the skyviper PR

**Pull request closed without merging**